script.js

```diff
@@ -1,60 +1,274 @@
+      tableCurrencyCol.appendChild(newColumnHeader);
+
+      const rows = document.querySelectorAll('.table-row-currency tr');
+      rows.forEach(row => {
+         row.appendChild(document.createElement("td"));
+      });  
+   }
+
+   let existingRow = Array.from(tableCurrencyRow.querySelectorAll("tr")).find(row => {
+      return row.querySelector("th") && row.querySelector("th").innerText === dataRow;
+   });
+
+   if (existingRow) {
+      const colIndex = Array.from(tableCurrencyCol.querySelectorAll("th")).findIndex(header => header.innerText === dataCol);
+      if (colIndex > -1) {
+         let cell = existingRow.querySelectorAll("td")[colIndex - 1]; // -1 because row header is the first element
+         cell.innerText = rate;
+      }
+   }
+}
+
+// Function to add a new currency rate
+function addRate() {
+   const currencyFrom = document.querySelector('.currencyName_from').value.toUpperCase().trim();
+   const currencyTo = document.querySelector('.currencyName_to').value.toUpperCase().trim();
+   const rate = document.getElementById('currency_rate').value;
+
+   const existingRate = arrRates.find(ratesObj => ratesObj.base === currencyFrom);
 
    if (existingRate) {
-      existingRate.rates[currencyTo] = rate; //update existing currency conversion with a new rate
+      if (existingRate.rates[currencyTo] !== undefined) {
+         alert(`This currency conversion rate already exists`);
+      } else {
+         existingRate.rates[currencyTo] = rate;
+         addDataColTableCurrency(currencyFrom, currencyTo, rate);
+         addNewDataCurrency(currencyTo, 'update-currency_to');
+         addNewDataCurrency(currencyTo, 'datalistOptions_convert_to');
+      }
    } else {
       arrRates.push({
          timestamp: Date.now(),
          base: currencyFrom,
-         date: date,
          rates: { [currencyTo]: rate }
       });
+      addDataRowTableCurrency(currencyFrom);
+      addDataColTableCurrency(currencyFrom, currencyTo, rate);
+      addNewDataCurrency(currencyFrom, 'update-currency_from');
+      addNewDataCurrency(currencyFrom, 'datalistOptions_convert_from');
+      addNewDataCurrency(currencyTo, 'update-currency_to');
+      addNewDataCurrency(currencyTo, 'datalistOptions_convert_to');
    }
-   console.log(arrRates) 
 }
 
-function getConvert(event) {
-   event.preventDefault();
-   const amount = +(document.getElementById('amount').value);
+// Function to update an existing currency rate
+function updateRate() {
+   const newRate = parseFloat(document.getElementById('update-currency_rate').value);
+   const updateCurrencyFrom = document.getElementById('update-currency_from').value;
+   const updateCurrencyTo = document.getElementById('update-currency_to').value;
+
+   let existingRate = arrRates.find(rateObj => rateObj.base === updateCurrencyFrom);
+   if (existingRate) {
+      existingRate.rates[updateCurrencyTo] = newRate;
+      addDataColTableCurrency(updateCurrencyFrom, updateCurrencyTo, newRate);
+   }
+}
```

Comment:
From a UX perspective then I would actually call this function when changes are made to the "From" and "To" currency selectors. As soon as I change currency it automatically updates the new currency conversion.

Also on the "amount" input field if string is empty don't updateRate() which results in a `NaN` result

Minor UX improvements